pontoon/pretranslation/utils.py

```diff
@@ -58,17 +64,29 @@ def pretranslate(project, locales=None, entities=None):
     index = -1
 
     for locale in locales:
+        log.info("Fetching pretranslations for locale {} started".format(locale.code))
         for entity in entities:
             locale_entity = '{}-{}'.format(locale.id, entity.id)
+            plural_forms = range(0, locale.nplurals or 1)
+
             if locale_entity not in translated_entities:
-                string = ""
+                strings = []
                 tm_response = get_translation_memory_data(
                     text=entity.string,
                     locale=locale,
                 )
 
-                if tm_response and int(tm_response[0]['quality']) == 100:
-                    string = tm_response[0]['target']
+                tm_response = [t for t in tm_response if int(t['quality']) == 100]
+                tm_response = sorted(tm_response, key=lambda e: e['plural_form'])
+
+                if tm_response:
+                    if entity.string_plural == "":
+                        strings = [(tm_response[0]['target'], None)]
+                    else:
+                        for plural_form in plural_forms:
+                            plurals = [t for t in tm_response if t['plural_form'] == plural_form]
+                            if plurals:
+                                strings.append(((plurals[0]['target'], plural_form)))
```

Comment:
I don't understand how this works. You only make one request to TM (with singular string as text input), but then you use it for plural translations, in case translation (of a singular string) is stored as pluralized.

I think you should simplify the algorithm and use the same logic for MT:
1. Get TM entry for the singular string (`text=entity.string`).
1. `if entity.string_plural != ""`, also get TM entry for the pluralized string (`text=entity.string_plural`).
1. Assign `#1` to the first `plural_form` of the locale.
1. Assign `#1` to all remaining `plural_forms` of the locale.

Reply:
As a general rule, when I have to write a complex algorithm like this one, I tend to start by writing comments describing what it does, then I break it down into functions. 

This here is very difficult to read, and thus difficult both to understand and to maintain. I highly recommend that you break it down into smaller bits of code. It's fine to have multiple small functions in your code. It will also make your code easier to reason about, and might help you address the issues Matjaz pointed.

Also, this is in `utils.py`. That name is a convention that I dislike, as it is very unclear what its content is. For this `pretranslate` function, I would actually put it in a file that tells exactly what it contains (calling the file `pretranslate.py` is fine). This way you know that all functions in that file are strictly related to the pretranslate work, and it will be a lot easier for developers to understand and reason about this module.